src/provider/dh_keymgmt.c

```diff
@@ -0,0 +1,2065 @@
+                                             OSSL_KEYMGMT_SELECT_ALL_PARAMETERS);
+        if (rc != 1) {
+            ibmca_debug_op_ctx(genctx,
+                               "ERROR: ibmca_import_from_fallback_pkey failed");
+            goto error;
+        }
+    } else if (genctx->key != NULL) {
+        /* Copy parameters from template key */
+        if (ibmca_keymgmt_dh_dup_params(genctx->key, key) != 1) {
+            ibmca_debug_op_ctx(genctx,
+                               "ERROR: ibmca_keymgmt_dh_dup_params failed");
+            goto error;
+        }
+    } else {
+        put_error_key(key, IBMCA_ERR_INTERNAL_ERROR,
+                      "Neither parmgen context nor template key available");
+        goto error;
+    }
+
+    if (genctx->dh.gen.selection & OSSL_KEYMGMT_SELECT_KEYPAIR) {
+        if (ibmca_keymgmt_dh_gen_libica(key, genctx->dh.gen.priv_len, osslcb,
+                                        cbarg) != 1) {
+            ibmca_debug_op_ctx(genctx,
+                               "ERROR: ibmca_keymgmt_dh_gen_libica failed");
+
+            if (ibmca_keymgmt_dh_gen_fallback(genctx, key,
+                                              osslcb, cbarg) != 1) {
+                ibmca_debug_op_ctx(genctx,
+                                   "ERROR: ibmca_keymgmt_dh_gen_fallback failed");
+                goto error;
+            }
+        }
+     }
+
+    goto out;
+
+error:
+    ibmca_keymgmt_free(key);
+    key = NULL;
+
+out:
+    if (pkey != NULL)
+        EVP_PKEY_free(pkey);
+
+    return key;
+}
+
+static int ibmca_keymgmt_dh_has(const void *vkey, int selection)
+{
+    const struct ibmca_key *key = vkey;
+    int ok = 1;
+
+    if (key == NULL)
+        return 0;
+
+    ibmca_debug_key(key, "key: %p selection: 0x%x", key, selection);
+
+    if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)
+        ok = ok && (key->dh.pub != NULL);
+
+    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)
+        ok = ok && (key->dh.priv != NULL);
+
+    if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0) {
+        ok = ok && (key->dh.ffc_params.p != NULL &&
+                    key->dh.ffc_params.g != NULL);
+        if (key->type == EVP_PKEY_DHX)
+            ok = ok && (key->dh.ffc_params.q != NULL);
+    }
+
+    ibmca_debug_key(key, "ok: %d", ok);
+
+    return ok;
+}
+
+static int ibmca_keymgmt_dh_match(const void *vkey1, const void *vkey2,
+                                  int selection)
+{
+    const struct ibmca_key *key1 = vkey1;
+    const struct ibmca_key *key2 = vkey2;
+    int ok = 1, checked = 0;
+
+    if (key1 == NULL || key2 == NULL)
+        return 0;
+
+    ibmca_debug_key(key1, "key1: %p key2: %p selection: 0x%x", key1, key2,
+                    selection);
+
+    if (ibmca_keymgmt_match(key1, key2) == 0)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
+        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```

Comment:
no, OSSL_KEYMGMT_SELECT_KEYPAIR is a mask out of 2 bits:
```
# define OSSL_KEYMGMT_SELECT_KEYPAIR            \
    ( OSSL_KEYMGMT_SELECT_PRIVATE_KEY | OSSL_KEYMGMT_SELECT_PUBLIC_KEY )
```
So it can happen that only one of the bit is on, but the other is off.